spec-tests/src/main/java/org/candlepin/spec/bootstrap/client/request/Request.java

```diff
@@ -0,0 +1,332 @@
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.spec.bootstrap.client.request;
+
+import org.candlepin.ApiException;
+import org.candlepin.Pair;
+import org.candlepin.spec.bootstrap.client.ApiClient;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import okhttp3.Call;
+
+
+
+/**
+ * The Request class provides a fluent-style builder API for creating generic requests. The Request
+ * instance needs an ApiClient instance to perform the actual work, and can be provided either
+ * directly during construction, or can be pulled automatically from a client-level ApiClient
+ * object.
+ */
+public class Request {
+
+    private static final Pattern PATH_PARAM_REGEX = Pattern.compile("\\{([^}]+)\\}");
+
+    private final org.candlepin.ApiClient client;
+
+    private String basePath;
+    private String endpoint;
+    private String method;
+    private String body;
+
+    private Map<String, String> pathParams;
+    private Map<String, String> headers;
+    private Map<String, List<String>> queryParams;
+
+
+    /**
+     * Builds a new Request instance which will use the specified ApiClient to perform the actual
+     * request upon execution.
+     *
+     * @param client
+     *  the org.candlepin.ApiClient instance to use to execute the request
+     *
+     * @throws IllegalArgumentException
+     *  if the provided client is null
+     */
+    public Request(org.candlepin.ApiClient client) {
+        if (client == null) {
+            throw new IllegalArgumentException("client is null");
+        }
+
+        this.client = client;
+
+        this.pathParams = new HashMap<>();
+        this.headers = new HashMap<>();
+        this.queryParams = new HashMap<>();
+    }
+
+    /**
+     * Builds a new Request instance which will use the org.candlepin.ApiClient backing the specified
+     * client ApiClient to execute the request.
+     *
+     * @param client
+     *  The client ApiClient instance to use to execute the request
+     *
+     * @throws IllegalArgumentException
+     *  if the provided client is null
+     *
+     * @return
+     *  a newly instantiated Request instance using the given ApiClient
+     */
+    public static Request from(ApiClient client) {
+        if (client == null) {
+            throw new IllegalArgumentException("client is null");
+        }
+
+        return new Request(client.getApiClient());
+    }
+
+    // TODO: add a .to method which accepts an endpoint enum which define the HTTP method and endpoint
```

Comment:
Is this TODO intentional? Just want to double check and make sure this wasn't a method that you wanted to add in this PR and just forgot to implement it.

Reply:
Intentional, but won't be done anytime soon. If there's a push for how I feel all of this should actually work [1], then we'll investigate how much effort it'd be to add it.

[1] Basically, the spec file should generate some enums which themselves implement a shared interface (ApiEndpoint?). Each enum would provide the basic info for the endpoint: path, method, and some help info in string form for a nice .toString output. If such a thing comes to exist, the code for the request would look like the following:

```
Request.from(client)
  .to(OwnerContent.LIST_CONTENT) // GET /owners/{owner_key}/content; would set both fields
  .setPathParam("owner_key", owner.getKey())
  ...
```

I'd want to sort out something to codify the path params somehow, but that's potentially non-trivial (enums with enums?).